packages/connection/src/class-nonce-handler.php

```diff
@@ -0,0 +1,206 @@
+
+	/**
+	 * How many nonces should be removed during each run of the runtime cleanup.
+	 * Can be modified using the filter `jetpack_connection_nonce_cleanup_runtime_limit`.
+	 */
+	const CLEANUP_RUNTIME_LIMIT = 10;
+
+	/**
+	 * How many nonces should be removed per batch during the `clean_all()` run.
+	 */
+	const CLEAN_ALL_LIMIT_PER_BATCH = 1000;
+
+	/**
+	 * Nonce lifetime in seconds.
+	 */
+	const LIFETIME = HOUR_IN_SECONDS;
+
+	/**
+	 * The nonces used during the request are stored here to keep them valid.
+	 *
+	 * @var array
+	 */
+	private static $nonces_used_this_request = array();
+
+	/**
+	 * Adds a used nonce to a list of known nonces.
+	 *
+	 * @param int    $timestamp the current request timestamp.
+	 * @param string $nonce the nonce value.
+	 * @param bool   $run_cleanup Whether to run the `cleanup_runtime()`.
+	 *
+	 * @return bool whether the nonce is unique or not.
+	 */
+	public static function add( $timestamp, $nonce, $run_cleanup = true ) {
+		global $wpdb;
+
+		if ( isset( static::$nonces_used_this_request[ "$timestamp:$nonce" ] ) ) {
+			return static::$nonces_used_this_request[ "$timestamp:$nonce" ];
+		}
+
+		// This should always have gone through Jetpack_Signature::sign_request() first to check $timestamp and $nonce.
+		$timestamp = (int) $timestamp;
+		$nonce     = esc_sql( $nonce );
+
+		// Raw query so we can avoid races: add_option will also update.
+		$show_errors = $wpdb->hide_errors();
+
+		// Running try...finally to make sure.
+		try {
+			$old_nonce = $wpdb->get_row(
+				$wpdb->prepare( "SELECT 1 FROM `$wpdb->options` WHERE option_name = %s", "jetpack_nonce_{$timestamp}_{$nonce}" )
+			);
+
+			if ( is_null( $old_nonce ) ) {
+				$return = (bool) $wpdb->query(
+					$wpdb->prepare(
+						"INSERT INTO `$wpdb->options` (`option_name`, `option_value`, `autoload`) VALUES (%s, %s, %s)",
+						"jetpack_nonce_{$timestamp}_{$nonce}",
+						time(),
+						'no'
+					)
+				);
+
+				/**
+				 * Use the filter to disable the nonce cleanup that happens at shutdown after adding a new nonce.
+				 *
+				 * @since 9.0.0
+				 *
+				 * @param int $limit How many old nonces to remove at shutdown.
+				 */
+				if ( apply_filters( 'jetpack_connection_add_nonce_cleanup', $run_cleanup ) ) {
+					add_action( 'shutdown', array( __CLASS__, 'clean_runtime' ) );
+				}
+			} else {
+				$return = false;
+			}
+		} finally {
+			$wpdb->show_errors( $show_errors );
+		}
+
+		static::$nonces_used_this_request[ "$timestamp:$nonce" ] = $return;
+
+		return $return;
+	}
+
+	/**
+	 * Removing [almost] all the nonces.
+	 * Capped at 1 million records to avoid breaking the site.
+	 *
+	 * @param int $cutoff_timestamp All nonces added before this timestamp will be removed.
+	 *
+	 * @return true
+	 */
+	public static function clean_all( $cutoff_timestamp = PHP_INT_MAX ) {
+		// phpcs:ignore Generic.CodeAnalysis.ForLoopWithTestFunctionCall.NotAllowed
+		for ( $start_time = time(), $i = 0; $i < 1000 && time() < $start_time + 20; ++$i ) {
```

Comment:
Since we have a time limit now, should we drop the arbitrary "1000 batches" limit?

Also, it's a microoptimization, but we could avoid the addition on each iteration of the loop with something like
```
for ( $end_time = time() + 20; time() < $end_time; ) {
```